UV package management and Snowflake logging extras

dominictarro opened the pull request:
Two-part PR

1. Transferred package management to [uv](https://docs.astral.sh/uv/).
2. Added a Snowflake logging method to an extras module.

## Snowflake Logging

Gives `prefect_snowflake.SnowflakeConnector` users the ability to log commands executed by the cursor if it fails or during debug. This is particularly useful for pipelines generating SQL on the fly.

```python
from prefecto.blocks import lazy_load
from prefecto.ext.snowflake import PrefectoSnowflakeConnector


class Blocks:
    SNOWFLAKE_BLOCK_NAME: str = "snowflake-block-name"

    @property
    @lazy_load("SNOWFLAKE_BLOCK_NAME")
    def snowflake(self) -> PrefectoSnowflakeConnector:
        """Snowflake connector block."""

blocks = Blocks()

with blocks.snowflake.get_connection().cursor() as cursor:
    cursor.execute(
        "SELECT * FROM table WHERE id = %(id)s AND secret = %(secret)s",
        params={"id": 123, "secret": "shhh"},
        obfuscate_params=["secret"],
        command_id="secret-selection",
    )
```

```
INFO - [secret-selection] Beginning command.
DEBUG - [secret-selection] Executing command:
SELECT * FROM table WHERE id = 123 AND secret = ****
INFO - [secret-selection] Command executed successfully.
```

dominictarro committed:
refactor: pivot to uv package manager

- update project def
- in workflows, install uv and install packages through it (cache slower than reinstall)
- init uv.lock with existing deps

dominictarro committed:
refactor: remove docstring requirement

dominictarro committed:
feat: command execution logging for snowflake connector block

- logs start/end of snowflake command
- optionally display or mask parameters in command
- display command on failure
- extended connector block for wrapping the execute method

dominictarro committed:
refactor: migrate extension modules to a higher extensions module

dominictarro committed:
docs: Touch up dosctrings and add the mkdocs page

dominictarro committed:
fix: formatted all files with pre-commit

ybryzTE commented on src/prefecto/ext/snowflake.py:
```diff
@@ -0,0 +1,620 @@
+        ):
+            e.add_note(f"Command failed\n{formatted_command_string}")
+        logger.error("Command failed.", exc_info=e, stack_info=True)
+        raise e
```
I think we can change it to this to avoid checking the versions:
```python
if hasattr(e, "add_note") and callable(e.add_note):
    e.add_note("Command failed."...)
```

ybryzTE commented on src/prefecto/ext/snowflake.py:
```diff
@@ -0,0 +1,620 @@
+        connection = self._credentials_get_client(**connect_kwargs, **connect_params)
+        self._connection = connection
+        self.logger.info("Started a new connection to Snowflake.")
+        return connection
```
Correct me if I'm wrong, but this one follows the same logic found in `prefect_snowflake`, just injecting the custom `PrefectoSnowflakeConnection` at the end of `_credentials_get_client`, whereas in `prefect_snowflake` it calls `snowflake.connector.connect(**connect_params)` instead. Are there other mechanisms that `prefect_snowflake` provides to accomplish this without rewriting the same logic? I'd rather use `PrefectoSnowflakeCursor` directly with the call to `cursor(PrefectoSnowflakeCursor)` on the original connection object returned from `SnowflakeConnector`. Otherwise, one possible approach that I can think of to remove the duplicate logic is to use composition over inheritance.

dominictarro replied:
I like that idea more. I removed the relationship to prefect-snowflake and focused on keeping this a cursor class like `DictCursor`. It raises the question of "why is this in prefecto?", but it will be used alongside Prefect enough to be worth it as an extension.

```py
from snowflake.connector import connect
from prefecto.ext.snowflake import LogCursor

with connect(...).cursor(LogCursor) as cursor:
    cursor.execute(
        "SELECT * FROM table WHERE id = %(id)s AND secret = %(secret)s",
        params={"id": 123, "secret": "shhh"},
        obfuscate_params=["secret"],
        command_id="secret-selection",
    )
```

```
INFO - [secret-selection] Beginning command.
DEBUG - [secret-selection] Executing command:
SELECT * FROM table WHERE id = 123 AND secret = ****
INFO - [secret-selection] Command executed successfully.
```

ybryzTE commented on src/prefecto/ext/snowflake.py:
```diff
@@ -0,0 +1,620 @@
+
+    def cursor(self) -> PrefectoSnowflakeCursor:
+        """Creates a wrapped cursor object. Each statement will be executed in a new cursor object."""
+        return super().cursor(PrefectoSnowflakeCursor)
```
Pass the `SnowflakeConnection` to the constructor instead:
```python
class PrefectoSnowflakeConnection:
    def __init__(self, conn: _SnowflakeConnection):
        self._conn = conn
```
For the `cursor` method use `self._conn` to create a `PrefectoSnowflakeCursor`. Proxy all other calls to the underlying `self._conn`.

dominictarro replied:
I removed the connection class.

ybryzTE commented on src/prefecto/ext/snowflake.py:
```diff
@@ -0,0 +1,620 @@
+
+        return PrefectoSnowflakeConnection(**connect_params)
+
+    def get_connection(self, **connect_kwargs: Any) -> PrefectoSnowflakeConnection:
```
In `get_connection` you would just have
```python
self._connection = PrefectoSnowflakeConnection(super().get_connection(**connect_kwargs))
```

dominictarro replied:
I removed the connection class.

ybryzTE reviewed (commented):
Looks good. Got couple questions and propostions before approving.

dominictarro committed:
refactor: simplify error handling logic in `_execute`

dominictarro committed:
refactor: Remove the block implementations and update docstrings

This approach emphasizes just using the Snowflake connection object's `.cursor(cursor_cls)` functionality.

dominictarro committed:
conf: update dependency requirements

dominictarro committed:
refactor: Move snowflake custom cursor unit test to extensions test dir

dominictarro committed:
docs: Add `LogCursor` class docstring

dominictarro committed:
docs: Fix snowflake ext module docstring

ybryzTE reviewed (approved):
Looks good!
